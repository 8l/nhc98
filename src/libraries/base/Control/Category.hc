#include "newmacros.h"
#include "runtime.h"

#define CT_v283	((void*)startLabel+32)
#define CT_v284	((void*)startLabel+68)
#define CT_v285	((void*)startLabel+108)
#define CT_v286	((void*)startLabel+144)
#define CT_v287	((void*)startLabel+196)
#define FN_LAMBDA280	((void*)startLabel+224)
#define CT_v289	((void*)startLabel+244)
#define CF_LAMBDA280	((void*)startLabel+252)
#define CT_v290	((void*)startLabel+280)
#define FN_LAMBDA281	((void*)startLabel+308)
#define CT_v292	((void*)startLabel+328)
#define CF_LAMBDA281	((void*)startLabel+336)
#define CT_v293	((void*)startLabel+360)
#define CT_v294	((void*)startLabel+392)
#define CT_v295	((void*)startLabel+432)
#define ST_v288	((void*)startLabel+452)
#define ST_v291	((void*)startLabel+493)
extern Node FN_Control_46Category_46_46[];
extern Node FN_Prelude_46_46[];
extern Node F0_Prelude_46id[];
extern Node FN_NHC_46Internal_46_95noMethodError[];

static Node startLabel[] = {
  bytes2word(3,0,2,1)
, bytes2word(1,2,0,3)
, useLabel(CT_v283)
,};
Node FN_Control_46Category_46_62_62_62[] = {
  bytes2word(NEEDHEAP_I32,PUSH_ARG_I2,PUSH_ARG_I3,PUSH_HEAP)
, bytes2word(HEAP_CVAL_I3,HEAP_ARG,1,ZAP_ARG_I1)
, bytes2word(ZAP_ARG_I2,ZAP_ARG_I3,EVAL,NEEDHEAP_I32)
, bytes2word(APPLY,2,RETURN_EVAL,ENDCODE)
, bytes2word(0,0,0,0)
,	/* CT_v283: (byte 0) */
  HW(1,3)
, 0
,};
Node F0_Control_46Category_46_62_62_62[] = {
  CAPTAG(useLabel(FN_Control_46Category_46_62_62_62),3)
, VAPTAG(useLabel(FN_Control_46Category_46_46))
, bytes2word(1,0,0,1)
, useLabel(CT_v284)
,};
Node FN_Control_46Category_46_60_60_60[] = {
  bytes2word(NEEDHEAP_I32,PUSH_HEAP,HEAP_CVAL_I3,HEAP_ARG)
, bytes2word(1,RETURN_EVAL,ENDCODE,0)
, bytes2word(0,0,0,0)
,	/* CT_v284: (byte 0) */
  HW(1,1)
, 0
,};
Node F0_Control_46Category_46_60_60_60[] = {
  CAPTAG(useLabel(FN_Control_46Category_46_60_60_60),1)
, VAPTAG(useLabel(FN_Control_46Category_46_46))
, bytes2word(0,0,2,0)
, bytes2word(1,1,0,2)
, useLabel(CT_v285)
,};
Node FN_Control_46Category_46Category_46Prelude_46_45_62_46_46[] = {
  bytes2word(NEEDHEAP_I32,PUSH_HEAP,HEAP_CVAL_I3,HEAP_ARG_ARG_RET_EVAL)
, bytes2word(1,2,ENDCODE,0)
, bytes2word(0,0,0,0)
,	/* CT_v285: (byte 0) */
  HW(1,2)
, 0
,};
Node F0_Control_46Category_46Category_46Prelude_46_45_62_46_46[] = {
  CAPTAG(useLabel(FN_Control_46Category_46Category_46Prelude_46_45_62_46_46),2)
, VAPTAG(useLabel(FN_Prelude_46_46))
, bytes2word(0,0,0,0)
, useLabel(CT_v286)
,};
Node FN_Control_46Category_46Category_46Prelude_46_45_62_46id[] = {
  bytes2word(NEEDSTACK_I16,PUSH_CVAL_P1,3,RETURN_EVAL)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v286: (byte 0) */
  HW(1,0)
, 0
,};
Node CF_Control_46Category_46Category_46Prelude_46_45_62_46id[] = {
  VAPTAG(useLabel(FN_Control_46Category_46Category_46Prelude_46_45_62_46id))
, useLabel(F0_Prelude_46id)
, bytes2word(3,0,2,1)
, bytes2word(1,2,0,3)
, useLabel(CT_v287)
,};
Node FN_Control_46Category_46_95_46_46[] = {
  bytes2word(ZAP_ARG_I1,NEEDHEAP_I32,PUSH_ARG_I3,PUSH_ARG_I2)
, bytes2word(HEAP_CVAL_I3,PUSH_HEAP,HEAP_CVAL_I4,HEAP_OFF_N1)
, bytes2word(2,ZAP_ARG_I2,ZAP_ARG_I3,EVAL)
, bytes2word(NEEDHEAP_I32,APPLY,2,RETURN_EVAL)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v287: (byte 0) */
  HW(2,3)
, 0
,};
Node F0_Control_46Category_46_95_46_46[] = {
  CAPTAG(useLabel(FN_Control_46Category_46_95_46_46),3)
, VAPTAG(useLabel(FN_LAMBDA280))
, VAPTAG(useLabel(FN_NHC_46Internal_46_95noMethodError))
, bytes2word(0,0,0,0)
, useLabel(CT_v289)
,	/* FN_LAMBDA280: (byte 0) */
  bytes2word(NEEDHEAP_I32,PUSH_CADR_N1,2,STRING)
, bytes2word(RETURN,ENDCODE,0,0)
, bytes2word(0,0,0,0)
, CONSTRW(1,0)
, useLabel(ST_v288)
,	/* CT_v289: (byte 0) */
  HW(0,0)
, 0
,	/* CF_LAMBDA280: (byte 0) */
  VAPTAG(useLabel(FN_LAMBDA280))
, bytes2word(1,0,0,1)
, useLabel(CT_v290)
,};
Node FN_Control_46Category_46_95_46id[] = {
  bytes2word(ZAP_ARG_I1,NEEDHEAP_I32,HEAP_CVAL_I3,PUSH_HEAP)
, bytes2word(HEAP_CVAL_I4,HEAP_OFF_N1,2,RETURN_EVAL)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v290: (byte 0) */
  HW(2,1)
, 0
,};
Node F0_Control_46Category_46_95_46id[] = {
  CAPTAG(useLabel(FN_Control_46Category_46_95_46id),1)
, VAPTAG(useLabel(FN_LAMBDA281))
, VAPTAG(useLabel(FN_NHC_46Internal_46_95noMethodError))
, bytes2word(0,0,0,0)
, useLabel(CT_v292)
,	/* FN_LAMBDA281: (byte 0) */
  bytes2word(NEEDHEAP_I32,PUSH_CADR_N1,2,STRING)
, bytes2word(RETURN,ENDCODE,0,0)
, bytes2word(0,0,0,0)
, CONSTRW(1,0)
, useLabel(ST_v291)
,	/* CT_v292: (byte 0) */
  HW(0,0)
, 0
,	/* CF_LAMBDA281: (byte 0) */
  VAPTAG(useLabel(FN_LAMBDA281))
, bytes2word(1,0,0,1)
, useLabel(CT_v293)
,};
Node FN_Control_46Category_46_46[] = {
  bytes2word(NEEDSTACK_I16,SELECTOR_EVAL,SELECT,1)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v293: (byte 0) */
  HW(0,1)
, 0
,};
Node F0_Control_46Category_46_46[] = {
  CAPTAG(useLabel(FN_Control_46Category_46_46),1)
, bytes2word(1,0,0,1)
, useLabel(CT_v294)
,};
Node FN_Control_46Category_46id[] = {
  bytes2word(NEEDSTACK_I16,SELECTOR_EVAL,SELECT,2)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v294: (byte 0) */
  HW(0,1)
, 0
,};
Node F0_Control_46Category_46id[] = {
  CAPTAG(useLabel(FN_Control_46Category_46id),1)
, bytes2word(0,0,0,0)
, useLabel(CT_v295)
,};
Node FN_Control_46Category_46Category_46Prelude_46_45_62[] = {
  bytes2word(NEEDHEAP_I32,PUSH_HEAP,HEAP_CVAL_N1,1)
, bytes2word(HEAP_CVAL_I3,HEAP_CVAL_I4,RETURN,ENDCODE)
, bytes2word(0,0,0,0)
, 0
, CONSTR(0,2,0)
,	/* CT_v295: (byte 0) */
  HW(2,0)
, 0
,};
Node CF_Control_46Category_46Category_46Prelude_46_45_62[] = {
  VAPTAG(useLabel(FN_Control_46Category_46Category_46Prelude_46_45_62))
, useLabel(F0_Control_46Category_46Category_46Prelude_46_45_62_46_46)
, useLabel(CF_Control_46Category_46Category_46Prelude_46_45_62_46id)
,	/* ST_v288: (byte 0) */
  bytes2word(78,111,32,100)
, bytes2word(101,102,97,117)
, bytes2word(108,116,32,100)
, bytes2word(101,102,105,110)
, bytes2word(105,116,105,111)
, bytes2word(110,32,102,111)
, bytes2word(114,32,99,108)
, bytes2word(97,115,115,32)
, bytes2word(109,101,116,104)
, bytes2word(111,100,32,46)
,	/* ST_v291: (byte 1) */
  bytes2word(0,78,111,32)
, bytes2word(100,101,102,97)
, bytes2word(117,108,116,32)
, bytes2word(100,101,102,105)
, bytes2word(110,105,116,105)
, bytes2word(111,110,32,102)
, bytes2word(111,114,32,99)
, bytes2word(108,97,115,115)
, bytes2word(32,109,101,116)
, bytes2word(104,111,100,32)
, bytes2word(105,100,0,0)
,};
