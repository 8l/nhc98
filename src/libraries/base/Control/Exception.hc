#include "newmacros.h"
#include "runtime.h"

#define FN_Control_46Exception_46catchesHandler	((void*)startLabel+16)
#define CT_v261	((void*)startLabel+44)
#define F0_Control_46Exception_46catchesHandler	((void*)startLabel+52)
#define FN_Control_46Exception_46Prelude_46214_46tryHandler	((void*)startLabel+88)
#define v265	((void*)startLabel+110)
#define v266	((void*)startLabel+113)
#define CT_v269	((void*)startLabel+136)
#define F0_Control_46Exception_46Prelude_46214_46tryHandler	((void*)startLabel+144)
#define CT_v270	((void*)startLabel+184)
extern Node FN_Control_46Exception_46Base_46throw[];
extern Node CF_Control_46Exception_46Base_46Exception_46Control_46Exception_46Base_46SomeException[];
extern Node FN_NHC_46Internal_46_95apply1[];
extern Node FN_Prelude_46foldr[];
extern Node FN_Control_46Exception_46Base_46fromException[];
extern Node FN_Control_46Exception_46Base_46catch[];

static Node startLabel[] = {
  42
, bytes2word(0,0,2,0)
, bytes2word(1,1,0,2)
, useLabel(CT_v261)
,	/* FN_Control_46Exception_46catchesHandler: (byte 0) */
  bytes2word(NEEDHEAP_I32,HEAP_CVAL_I3,HEAP_ARG,2)
, bytes2word(HEAP_CVAL_I4,HEAP_CVAL_I5,HEAP_CVAL_P1,6)
, bytes2word(HEAP_OFF_N1,3,HEAP_ARG,2)
, bytes2word(PUSH_HEAP,HEAP_CVAL_P1,7,HEAP_OFF_N1)
, bytes2word(8,HEAP_OFF_N1,5,HEAP_ARG)
, bytes2word(1,RETURN_EVAL,ENDCODE,0)
, bytes2word(0,0,0,0)
,	/* CT_v261: (byte 0) */
  HW(5,2)
, 0
,	/* F0_Control_46Exception_46catchesHandler: (byte 0) */
  CAPTAG(useLabel(FN_Control_46Exception_46catchesHandler),2)
, CAPTAG(useLabel(FN_Control_46Exception_46Prelude_46214_46tryHandler),2)
, VAPTAG(useLabel(FN_Control_46Exception_46Base_46throw))
, useLabel(CF_Control_46Exception_46Base_46Exception_46Control_46Exception_46Base_46SomeException)
, VAPTAG(useLabel(FN_NHC_46Internal_46_95apply1))
, VAPTAG(useLabel(FN_Prelude_46foldr))
, bytes2word(3,0,2,1)
, bytes2word(1,2,0,3)
, useLabel(CT_v269)
,	/* FN_Control_46Exception_46Prelude_46214_46tryHandler: (byte 0) */
  bytes2word(NEEDSTACK_I16,PUSH_ZAP_ARG_I2,EVAL,NEEDHEAP_I32)
, bytes2word(UNPACK,2,PUSH_ARG_I1,PUSH_HEAP)
, bytes2word(HEAP_CVAL_I3,HEAP_I2,ZAP_ARG_I1,EVAL)
, bytes2word(NEEDHEAP_I32,APPLY,1,EVAL)
, bytes2word(TABLESWITCH,2,TOP(4),BOT(4))
,	/* v265: (byte 2) */
  bytes2word(TOP(7),BOT(7),POP_I1,PUSH_ARG_I3)
,	/* v266: (byte 1) */
  bytes2word(RETURN_EVAL,UNPACK,1,PUSH_P1)
, bytes2word(0,PUSH_P1,3,ZAP_ARG_I3)
, bytes2word(ZAP_STACK_P1,4,ZAP_STACK_P1,3)
, bytes2word(ZAP_STACK_P1,2,EVAL,NEEDHEAP_I32)
, bytes2word(APPLY,1,RETURN_EVAL,ENDCODE)
, bytes2word(0,0,0,0)
,	/* CT_v269: (byte 0) */
  HW(1,3)
, 0
,	/* F0_Control_46Exception_46Prelude_46214_46tryHandler: (byte 0) */
  CAPTAG(useLabel(FN_Control_46Exception_46Prelude_46214_46tryHandler),3)
, VAPTAG(useLabel(FN_Control_46Exception_46Base_46fromException))
, bytes2word(0,0,2,0)
, bytes2word(1,1,0,2)
, useLabel(CT_v270)
,};
Node FN_Control_46Exception_46catches[] = {
  bytes2word(NEEDHEAP_I32,HEAP_CVAL_I3,HEAP_ARG,2)
, bytes2word(PUSH_HEAP,HEAP_CVAL_I4,HEAP_CVAL_I5,HEAP_ARG)
, bytes2word(1,HEAP_OFF_N1,5,RETURN_EVAL)
, bytes2word(ENDCODE,0,0,0)
, bytes2word(0,0,0,0)
,	/* CT_v270: (byte 0) */
  HW(3,2)
, 0
,};
Node F0_Control_46Exception_46catches[] = {
  CAPTAG(useLabel(FN_Control_46Exception_46catches),2)
, CAPTAG(useLabel(FN_Control_46Exception_46catchesHandler),1)
, VAPTAG(useLabel(FN_Control_46Exception_46Base_46catch))
, useLabel(CF_Control_46Exception_46Base_46Exception_46Control_46Exception_46Base_46SomeException)
,};
