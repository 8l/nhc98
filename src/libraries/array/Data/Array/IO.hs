{-# OPTIONS_GHC -#include "HsBase.h" #-}
{-# OPTIONS_GHC -w #-} --tmp
-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Array.IO
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
-- 
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  non-portable (uses Data.Array.MArray)
--
-- Mutable boxed and unboxed arrays in the IO monad.
--
-----------------------------------------------------------------------------

module Data.Array.IO (
   -- * @IO@ arrays with boxed elements
   IOArray,		-- instance of: Eq, Typeable

   -- * @IO@ arrays with unboxed elements
   IOUArray,		-- instance of: Eq, Typeable
   castIOUArray,	-- :: IOUArray i a -> IO (IOUArray i b)

   -- * Overloaded mutable array interface
   module Data.Array.MArray,

   -- * Doing I\/O with @IOUArray@s
   hGetArray,		-- :: Handle -> IOUArray Int Word8 -> Int -> IO Int
   hPutArray,		-- :: Handle -> IOUArray Int Word8 -> Int -> IO ()
 ) where

import Data.Array.Base
import Data.Array.IO.Internals
import Data.Array.MArray
import System.IO.Error

#ifdef __GLASGOW_HASKELL__
import Foreign
import Foreign.C

import GHC.Exts  (MutableByteArray#, RealWorld)
import GHC.Arr
import GHC.IORef
import GHC.IO.Handle
import GHC.IO.Buffer
import GHC.IO.Exception

#else
import Data.Char
import Data.Word ( Word8 )
import System.IO
#endif

#ifdef __GLASGOW_HASKELL__
-- ---------------------------------------------------------------------------
-- hGetArray

-- | Reads a number of 'Word8's from the specified 'Handle' directly
-- into an array.
hGetArray
 	:: Handle		-- ^ Handle to read from
	-> IOUArray Int Word8	-- ^ Array in which to place the values
	-> Int			-- ^ Number of 'Word8's to read
	-> IO Int
		-- ^ Returns: the number of 'Word8's actually 
		-- read, which might be smaller than the number requested
		-- if the end of file was reached.

hGetArray handle (IOUArray (STUArray _l _u n ptr)) count
  | count == 0              = return 0
  | count < 0 || count > n  = illegalBufferSize handle "hGetArray" count
  | otherwise = do
      -- we would like to read directly into the buffer, but we can't
      -- be sure that the MutableByteArray# is pinned, so we have to
      -- allocate a separate area of memory and copy.
      allocaBytes n $ \p -> do
        r <- hGetBuf handle p n
        memcpy_ba_ptr ptr p (fromIntegral r)
        return r

foreign import ccall unsafe "memcpy"
   memcpy_ba_ptr :: MutableByteArray# RealWorld -> Ptr a -> CSize -> IO (Ptr ())

-- ---------------------------------------------------------------------------
-- hPutArray

-- | Writes an array of 'Word8' to the specified 'Handle'.
hPutArray
	:: Handle			-- ^ Handle to write to
	-> IOUArray Int Word8		-- ^ Array to write from
	-> Int				-- ^ Number of 'Word8's to write
	-> IO ()

hPutArray handle (IOUArray (STUArray _l _u n raw)) count
  | count == 0              = return ()
  | count < 0 || count > n  = illegalBufferSize handle "hPutArray" count
  | otherwise = do
      -- as in hGetArray, we would like to use the array directly, but
      -- we can't be sure that the MutableByteArray# is pinned.
     allocaBytes n $ \p -> do
       memcpy_ptr_ba p raw (fromIntegral n)
       hPutBuf handle p n

foreign import ccall unsafe "memcpy"
   memcpy_ptr_ba :: Ptr a -> MutableByteArray# RealWorld -> CSize -> IO (Ptr ())

-- ---------------------------------------------------------------------------
-- Internal Utils

illegalBufferSize :: Handle -> String -> Int -> IO a
illegalBufferSize handle fn sz = 
	ioException (ioeSetErrorString
		     (mkIOError InvalidArgument fn (Just handle) Nothing)
		     ("illegal buffer size " ++ showsPrec 9 (sz::Int) []))

#else /* !__GLASGOW_HASKELL__ */
hGetArray :: Handle -> IOUArray Int Word8 -> Int -> IO Int
hGetArray handle arr count = do
	bds <- getBounds arr
	if count < 0 || count > rangeSize bds
	   then illegalBufferSize handle "hGetArray" count
	   else get 0
 where
  get i | i == count = return i
	| otherwise = do
		error_or_c <- try (hGetChar handle)
		case error_or_c of
		    Left ex
			| isEOFError ex -> return i
			| otherwise -> ioError ex
		    Right c -> do
			unsafeWrite arr i (fromIntegral (ord c))
			get (i+1)

hPutArray :: Handle -> IOUArray Int Word8 -> Int -> IO ()
hPutArray handle arr count = do
	bds <- getBounds arr
	if count < 0 || count > rangeSize bds
	   then illegalBufferSize handle "hPutArray" count
	   else put 0
 where
  put i | i == count = return ()
	| otherwise = do
		w <- unsafeRead arr i
		hPutChar handle (chr (fromIntegral w))
		put (i+1)

illegalBufferSize :: Handle -> String -> Int -> IO a
illegalBufferSize _ fn sz = ioError $
	userError (fn ++ ": illegal buffer size " ++ showsPrec 9 (sz::Int) [])
#endif /* !__GLASGOW_HASKELL__ */
