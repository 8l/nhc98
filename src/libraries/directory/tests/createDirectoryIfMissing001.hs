module Main(main) where

import Control.Concurrent
import Control.Monad
import Control.Exception
import System.Directory
import System.FilePath
import System.IO.Error hiding (try)

testdir = "createDirectoryIfMissing001.d"
testdir_a = testdir </> "a"

main = do
  cleanup

  report $ createDirectoryIfMissing False testdir
  cleanup

  report $ createDirectoryIfMissing False testdir_a
   -- should fail with does not exist

  report $ createDirectoryIfMissing True testdir_a
   -- should succeed with no error
  report $ createDirectoryIfMissing False testdir_a
   -- should succeed with no error
  report $ createDirectoryIfMissing False (addTrailingPathSeparator testdir_a)
   -- should succeed with no error

  cleanup
  report $ createDirectoryIfMissing True (addTrailingPathSeparator testdir_a)

  -- look for race conditions: #2808.  This fails with
  -- +RTS -N2 and directory 1.0.0.2.
  m <- newEmptyMVar
  forkIO $ do replicateM_ 10000 create; putMVar m ()
  forkIO $ do replicateM_ 10000 cleanup; putMVar m ()
  replicateM_ 2 $ takeMVar m

-- This test fails on Windows; see #2924
--  replicateM_ 2 $ 
--     forkIO $ do replicateM_ 5000 (do create; cleanup); putMVar m ()
--  replicateM_ 2 $ takeMVar m

  cleanup

-- createDirectoryIfMissing is allowed to fail with isDoesNotExistError if
-- another process/thread removes one of the directories during the proces
-- of creating the hierarchy.
create = tryJust (guard . isDoesNotExistError) $ createDirectoryIfMissing True testdir_a

cleanup = ignore $ removeDirectoryRecursive testdir

report :: Show a => IO a -> IO ()
report io = do
  r <- try io
  case r of
   Left e  -> print (e :: SomeException)
   Right a -> print a

ignore :: IO a -> IO ()
ignore io = do
  r <- try io
  case r of
   Left e  -> let _ = e :: SomeException in return ()
   Right a -> return ()
