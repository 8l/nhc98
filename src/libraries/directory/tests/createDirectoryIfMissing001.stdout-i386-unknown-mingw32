()
CreateDirectory "createDirectoryIfMissing001.d\\a": does not exist (The system cannot find the path specified.)
()
()
()
()
