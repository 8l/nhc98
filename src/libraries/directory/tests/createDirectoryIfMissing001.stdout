()
createDirectoryIfMissing001.d/a: createDirectory: does not exist (No such file or directory)
()
()
()
()
